src/services/datasource.ts

```diff
@@ -32,13 +39,29 @@ type IndexVolumeResponse = {
   };
 };
 
+type SampleTimeStamp = number;
+type SampleCount = number;
+type PatternSample = [SampleTimeStamp, SampleCount];
+
+export interface LokiPattern {
+  pattern: string;
+  samples: PatternSample[];
+}
+
+type PatternsResponse = {
+  data: LokiPattern[];
+};
+
 class WrappedLokiDatasource extends RuntimeDataSource<DataQuery> {
   constructor(pluginId: string, uid: string) {
     super(pluginId, uid);
   }
 
   query(request: SceneDataQueryRequest): Promise<DataQueryResponse> | Observable<DataQueryResponse> {
-    return new Observable<DataQueryResponse>((subscriber) => {
+    const numberOfQueries = request.targets.length;
+    let numberOfQueriesThatResolved = 0;
+
+    const observable = new Observable<DataQueryResponse>((subscriber) => {
```

Comment:
clean this up too, be good to throw an error when a resource call makes it to the data request